src/providers/hana/qgshanaproviderconnection.cpp

```diff
@@ -0,0 +1,354 @@
+  }
+  catch ( const QgsHanaException &ex )
+  {
+    throw QgsProviderConnectionException( QString( ex.what() ) );
+  }
+
+  return results;
+}
+
+void QgsHanaProviderConnection::executeSqlStatement( const QString &sql ) const
+{
+  const QgsDataSourceUri dsUri { uri() };
+  QgsHanaConnectionRef conn( dsUri );
+  if ( conn.isNull() )
+    throw QgsProviderConnectionException( QObject::tr( "Connection failed: %1" ).arg( uri() ) );
+
+  try
+  {
+    ConnectionRef connRef = conn->getNativeRef();
+    StatementRef stmt = connRef->createStatement();
+    stmt->execute( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    connRef->commit();
+  }
+  catch ( const odbc::Exception &ex )
+  {
+    throw QgsProviderConnectionException( QString( ex.what() ) );
+  }
+}
+
+QList<QgsHanaProviderConnection::TableProperty> QgsHanaProviderConnection::tables( const QString &schema, const TableFlags &flags ) const
+{
+  checkCapability( Capability::Tables );
+  QList<QgsHanaProviderConnection::TableProperty> tables;
+  const QgsDataSourceUri dsUri { uri() };
+
+  QgsHanaConnectionRef conn( dsUri );
+  if ( conn.isNull() )
+    throw QgsProviderConnectionException( QObject::tr( "Connection failed: %1" ).arg( uri() ) );
+
+  QVector<QgsHanaLayerProperty> layers;
+  try
+  {
+    layers = conn->getLayers( schema, flags.testFlag( TableFlag::Aspatial ), false );
+
+    for ( auto &layerInfo : layers )
+      conn->readLayerInfo( layerInfo );
+
+    for ( const auto &layerInfo : layers )
+    {
+      // Classify
+      TableFlags prFlags;
+      if ( layerInfo.isView )
+        prFlags.setFlag( QgsHanaProviderConnection::TableFlag::View );
+      else if ( !layerInfo.geometryColName.isEmpty() )
+        prFlags.setFlag( QgsHanaProviderConnection::TableFlag::Vector );
+      else
+        prFlags.setFlag( QgsHanaProviderConnection::TableFlag::Aspatial );
+
+      // Filter
+      if ( ! flags || ( prFlags & flags ) )
+      {
+        QgsHanaProviderConnection::TableProperty property;
+        property.setFlags( prFlags );
+
+        QgsCoordinateReferenceSystem crs = conn->getCrs( layerInfo.srid );
+        property.addGeometryColumnType( layerInfo.type, crs );
+
+        property.setTableName( layerInfo.tableName );
+        property.setSchema( layerInfo.schemaName );
+        property.setGeometryColumn( layerInfo.geometryColName );
+        property.setPrimaryKeyColumns( layerInfo.pkCols );
+        property.setGeometryColumnCount( layerInfo.geometryColName.isEmpty() ? 0 : 1 );
+        property.setComment( layerInfo.tableComment );
+        tables.push_back( property );
+      }
+    }
+  }
+  catch ( const odbc::Exception &ex )
+  {
+    throw QgsProviderConnectionException( QObject::tr( "Could not retrieve tables: %1, %2" ).arg( uri(), QString( ex.what() ) ) );
+  }
+  catch ( const QgsHanaException &ex )
+  {
+    throw QgsProviderConnectionException( QObject::tr( "Could not retrieve tables: %1, %2" ).arg( uri(), QString( ex.what() ) ) );
+  }
```

Comment:
This is again a not so nice result of mixing usage of native and wrapper objects.

Reply:
Done.